dali/pipeline/operators/arithmetic/arithmetic.h

```diff
@@ -0,0 +1,249 @@
+}
+
+
+struct ExpressionImplTask {
+  ExpressionImplBase * impl;
+  ExpressionImplContext ctx;
+};
+
+template <typename Backend>
+inline void CreateExecutionOrder(std::vector<ExpressionImplTask> &order, const ExprNode &expr,
+                                 ExprImplCache &cache) {
+  if (expr.GetNodeType() != NodeType::Function) {
+    return;
+  }
+  for (int i = 0; i < expr.GetSubexpressionCount(); i++) {
+    CreateExecutionOrder<Backend>(order, expr[i], cache);
+  }
+  order.push_back({cache.GetExprImpl<Backend>(expr), {&expr}});
+}
+
+template <typename Backend>
+inline std::vector<ExpressionImplTask> CreateExecutionOrder(const ExprNode &expr,
+                                                              ExprImplCache &cache) {
+  std::vector<ExpressionImplTask> result;
+  CreateExecutionOrder<Backend>(result, expr, cache);
+  return result;
+}
+
+inline kernels::TensorListShape<> ShapePromotion(std::string op,
+                                                 const kernels::TensorListShape<> &left,
+                                                 const kernels::TensorListShape<> &right) {
+  bool is_left_scalar = IsScalarLike(left);
+  bool is_right_scalar = IsScalarLike(right);
+  if (is_left_scalar && is_right_scalar) {
+    return kernels::TensorListShape<>{{1}};
+  }
+  if (is_left_scalar) {
+    return right;
+  }
+  if (is_right_scalar) {
+    return left;
+  }
+  using std::to_string;
+  DALI_ENFORCE(left == right, "Input shapes of elemenetwise arithemtic operator \"" + op +
+                                  "\" do not match. Expected equal shapes, got: " + op + "(" +
+                                  to_string(left) + ", " + to_string(right) + ").");
+  return left;
+}
+
+template <typename Backend>
+DLL_PUBLIC kernels::TensorListShape<> PropagateShapes(ExprNode &expr,
+                                                      const workspace_t<Backend> &ws) {
+  if (expr.GetNodeType() == NodeType::Constant) {
+    expr.SetShape(kernels::TensorListShape<>{{1}});
+    return expr.GetShape();
+  }
+  if (expr.GetNodeType() == NodeType::Tensor) {
+    auto &e = dynamic_cast<ExprTensor&>(expr);
+    expr.SetShape(ws.template InputRef<Backend>(e.GetMappedInput()).shape());
+    return expr.GetShape();
+  }
+  if (expr.GetSubexpressionCount() == 2) {
+    auto left_shape = PropagateShapes<Backend>(expr[0], ws);
+    auto right_shape = PropagateShapes<Backend>(expr[1], ws);
+    expr.SetShape(ShapePromotion(expr.GetOp(), left_shape, right_shape));
+    return expr.GetShape();
+  }
+  DALI_FAIL("Only binary expressions are supported");
+}
+
+template <typename Backend>
+class ArithmeticGenericOp : public Operator<Backend> {
+ public:
+  inline explicit ArithmeticGenericOp(const OpSpec &spec) : Operator<Backend>(spec) {
+    expr_ = ParseExpressionString(spec.GetArgument<std::string>("expression_desc"));
+    num_tasks_ = std::is_same<Backend, CPUBackend>::value ? 2 * num_threads_ : 1;
```

Comment:
Why `num_tasks_` is `2 * num_threads_`?

Reply:
To allow better distribution of tasks. I can make it slightly bigger, to `3` or `4`.

Reply:
I had a rare case when `num_tasks_ = num_threads_`, when one thread picked two tasks.

Reply:
ok